File: CPP/Lesson_059_time_based_key_value_store.cpp
// =============================================================
// MIT License | @analyticswithharry2026
// GitHub  : https://github.com/analyticswithharry
// YouTube : Analytics with Harry
// =============================================================
// Lesson     : 059 -- Time Based Key Value Store
// Category   : Binary Search
// Difficulty : Medium
// Study Plan : Day 29
// =============================================================
//
// QUESTION:
//   Design a time-based key-value data structure that supports
//   set(key, value, timestamp) and get(key, timestamp), where get returns
//   the value with the largest timestamp <= the given timestamp.
// =============================================================

#include <vector>
#include <string>
#include <iostream>
#include <stack>
#include <queue>
#include <unordered_map>
#include <unordered_set>
#include <map>
#include <set>
#include <algorithm>
#include <climits>
#include <numeric>
#include <functional>
#include <cmath>
using namespace std;
class TimeMap {
    unordered_map<string, vector<pair<int,string>>> d;
public:
    void set(string k, string v, int t){ d[k].push_back({t, v}); }
    string get(string k, int t){
        if (!d.count(k)) return "";
        auto& a = d[k];
        int l=0, r=a.size()-1; string res = "";
        while (l<=r) {
            int m = (l+r)/2;
            if (a[m].first <= t) { res = a[m].second; l = m+1; } else r = m-1;
        }
        return res;
    }
};
int main(){ TimeMap tm; tm.set("foo","bar",1);
    cout<<tm.get("foo",1)<<" "<<tm.get("foo",3)<<endl;
    tm.set("foo","bar2",4);
    cout<<tm.get("foo",4)<<" "<<tm.get("foo",5)<<endl; }

File: Java/Lesson059_TimeBasedKeyValueStore.java
// =============================================================
// MIT License | @analyticswithharry2026
// GitHub  : https://github.com/analyticswithharry
// YouTube : Analytics with Harry
// =============================================================
// Lesson     : 059 -- Time Based Key Value Store
// Category   : Binary Search
// Difficulty : Medium
// Study Plan : Day 29
// =============================================================
//
// QUESTION:
//   Design a time-based key-value data structure that supports
//   set(key, value, timestamp) and get(key, timestamp), where get returns
//   the value with the largest timestamp <= the given timestamp.
// =============================================================

import java.util.*;
public class Lesson059_TimeBasedKeyValueStore {
    static class TimeMap {
        Map<String, List<int[]>> ts = new HashMap<>();
        Map<String, List<String>> vs = new HashMap<>();
        public void set(String k, String v, int t){
            ts.computeIfAbsent(k, x -> new ArrayList<>()).add(new int[]{t});
            vs.computeIfAbsent(k, x -> new ArrayList<>()).add(v);
        }
        public String get(String k, int t){
            if (!ts.containsKey(k)) return "";
            List<int[]> a = ts.get(k); List<String> b = vs.get(k);
            int l = 0, r = a.size()-1; String res = "";
            while (l <= r) {
                int m = (l+r)/2;
                if (a.get(m)[0] <= t) { res = b.get(m); l = m+1; } else r = m-1;
            }
            return res;
        }
    }
    public static void main(String[] a){
        TimeMap tm = new TimeMap(); tm.set("foo","bar",1);
        System.out.println(tm.get("foo",1)); System.out.println(tm.get("foo",3));
        tm.set("foo","bar2",4);
        System.out.println(tm.get("foo",4)); System.out.println(tm.get("foo",5));
    }
}
